# Extracción por API de datos meteorologicos a AEMET

Un Input de gran valor para alimentar el modelo son las variables meteorologicas, ya que la relación de variables como irradiacion, nubosidad, temperatura, particularidad del aire o precipitaciones, estan intimamente ligadas a la producción solar.

Para ello se solicita la key a **AEMET** que se configurará en el siguiente.
En este caso la nuestra es:
> eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJqZXN1cy52aWxsYS5hbHZhcmV6QGdtYWlsLmNvbSIsImp0aSI6ImVkM2JkN2M0LWU4NDctNDQwZi04ODc2LTlhNmI2YTM1NWY1MCIsImlzcyI6IkFFTUVUIiwiaWF0IjoxNzAxMjU4MTU3LCJ1c2VySWQiOiJlZDNiZDdjNC1lODQ3LTQ0MGYtODg3Ni05YTZiNmEzNTVmNTAiLCJyb2xlIjoiIn0.52akELBtqCXQERxeeoORUzprT8dMbMEHOaKLtjiiiSk

## 1.Importado de librerias

In [63]:
#pip install requests
import requests
import csv
from warnings import warn

## 2.Solicitudes a la API de AEMET

In [70]:

api_key = 'eyJhbGciOiJIUzI1NiJ9.eyJzdWIiOiJqZXN1cy52aWxsYS5hbHZhcmV6QGdtYWlsLmNvbSIsImp0aSI6ImVkM2JkN2M0LWU4NDctNDQwZi04ODc2LTlhNmI2YTM1NWY1MCIsImlzcyI6IkFFTUVUIiwiaWF0IjoxNzAxMjU4MTU3LCJ1c2VySWQiOiJlZDNiZDdjNC1lODQ3LTQ0MGYtODg3Ni05YTZiNmEzNTVmNTAiLCJyb2xlIjoiIn0.52akELBtqCXQERxeeoORUzprT8dMbMEHOaKLtjiiiSk'
endpoint = 'https://opendata.aemet.es/opendata/api/avisos_cap/ultimoelaborado/area/61'

# Crea la URL de la solicitud con tu clave de API
url = f'{endpoint}?api_key={api_key}'

# Realiza la solicitud a la API
response = requests.get(url)
content = response.json()

#r_meta = requests.get(content['metadatos'], verify=False)
#r_data = requests.get(content['datos'], verify=False)

r_data = requests.get(content['datos'], verify=False)
if r_data.status_code == requests.codes.ok:
  data = r_data.json()
  prediction = data[0]['prediccion']
elif response.status_code == requests.codes.TOO_MANY_REQUESTS:
        warn('Too many requests')

elif response.status.code == requests.codes.NOT_FOUND:
        warn('No data for the response')

else:
        warn('Code error {}'.format(response.status_code))
print(prediction)



## WEBs de calidad del AIRE


La calidad del aire y el nivel de partículas en suspensión son medidas comunes que se proporcionan a través de diversas APIs. Algunas fuentes confiables y populares para obtener datos sobre la calidad del aire incluyen:



1.   **World Air Quality Index (WAQI):**
Sitio web: World Air Quality Index
API: WAQI proporciona una API que puedes utilizar para obtener datos sobre la calidad del aire en todo el mundo.
Endpoint de la API: https://api.waqi.info/feed/geo:{lat};{lon}/?token=YOUR_API_KEY
Necesitarás registrarte para obtener una clave de API.

2.   **OpenAQ:**
Sitio web: OpenAQ
API: OpenAQ ofrece una API que permite acceder a datos globales de calidad del aire.
Endpoint de la API: https://api.openaq.org/v1/measurements
Puedes consultar la documentación para más detalles y opciones.
AirNow API (para EE. UU.):

3.   **AirNow:**
Sitio web: AirNow
API: AirNow proporciona una API que ofrece información sobre la calidad del aire en los Estados Unidos.
Endpoint de la API: https://www.airnowapi.org/aq/observation/latLong/current/?format=application/json&latitude={lat}&longitude={lon}&distance=25&API_KEY=YOUR_API_KEY
Necesitarás registrarte y obtener una clave de API.

3.   **Plume Labs:**
Plume Labs - Plume API:
Sitio web: Plume Labs
API: Plume Labs ofrece una API que proporciona datos sobre la calidad del aire en varias ciudades del mundo.
Endpoint de la API: Puedes obtener más información sobre la API al ponerte en contacto con Plume Labs.

# https://aqicn.org/TOKEN:
ec8fe1cf76df3646ec4cffb994cdb529db513c68

# **World Air Quality Index (WAQI)**

In [77]:
import requests
import json

# URL de la API
url = "https://api.waqi.info/feed/geo:39.2234255;-5.4105358/?token=ec8fe1cf76df3646ec4cffb994cdb529db513c68"

# Realiza la solicitud a la API
response = requests.get(url)

# Verifica si la solicitud fue exitosa (código 200)
if response.status_code == 200:
    # Convierte la respuesta a formato JSON
    data = response.json()
    pretty_data = json.dumps(data, indent=2)
    print(pretty_data)
    # Accede a los datos que necesitas (puedes ajustar esto según la estructura de la respuesta)
    city_name = data['data']['city']['name']
    aqi = data['data']['aqi']

    # Imprime los resultados
    print(f"Calidad del aire en {city_name}: {aqi}")

    # Accede a los datos históricos o diarios (ajusta según la estructura real)
    datos_temporales = data['data']['forecast']['day']  # Esto es un ejemplo ficticio

    # Filtra los datos para la fecha deseada (ajusta según la estructura real)
    datos_para_fecha_deseada = [dato for dato in datos_temporales if dato['day'] == '2023-11-29']

    # Imprime los resultados para la fecha deseada
    if datos_para_fecha_deseada:
        print(json.dumps(datos_para_fecha_deseada[0], indent=2))
    else:
        print(f"No hay datos disponibles para la fecha {'2023-11-29'}")

else:
    # Si la solicitud no fue exitosa, imprime el código de estado
    print(f"Error en la solicitud. Código de estado: {response.status_code}")


{
  "status": "ok",
  "data": {
    "aqi": 56,
    "idx": 14175,
    "attributions": [
      {
        "url": "http://www.juntadeandalucia.es/medioambiente/site/portalweb/",
        "name": "Consejer\u00eda de Medio Ambiente y Ordenaci\u00f3n del Territorio :: Junta de Andaluc\u00eda",
        "logo": "Andalucia-Conserjeria-de-medio-ambiente.png"
      },
      {
        "url": "http://www.eea.europa.eu/themes/air/",
        "name": "European Environment Agency",
        "logo": "Europe-EEA.png"
      },
      {
        "url": "https://waqi.info/",
        "name": "World Air Quality Index Project"
      }
    ],
    "city": {
      "geo": [
        38.137927556412,
        -4.9039669638703
      ],
      "name": "Villaharta, Spain",
      "url": "https://aqicn.org/city/spain/andalucia/villaharta",
      "location": ""
    },
    "dominentpol": "pm25",
    "iaqi": {
      "dew": {
        "v": 14
      },
      "h": {
        "v": 93
      },
      "no2": {
        "v": 0.5
      },
   

KeyError: ignored

https://api.waqi.info/feed/geo:39.2234255;-5.4105358/?token=ec8fe1cf76df3646ec4cffb994cdb529db513c68

{"status":"ok","data":{"aqi":58,"idx":14175,"attributions":[{"url":"http://www.juntadeandalucia.es/medioambiente/site/portalweb/","name":"Consejería de Medio Ambiente y Ordenación del Territorio :: Junta de Andalucía","logo":"Andalucia-Conserjeria-de-medio-ambiente.png"},{"url":"http://www.eea.europa.eu/themes/air/","name":"European Environment Agency","logo":"Europe-EEA.png"},{"url":"https://waqi.info/","name":"World Air Quality Index Project"}],"city":{"geo":[38.137927556412,-4.9039669638703],"name":"Villaharta, Spain","url":"https://aqicn.org/city/spain/andalucia/villaharta","location":""},"dominentpol":"pm25","iaqi":{"dew":{"v":14},"h":{"v":100},"no2":{"v":0.5},"p":{"v":1010},"pm10":{"v":9},"pm25":{"v":58},"so2":{"v":2.1},"t":{"v":14},"w":{"v":2.8}},"time":{"s":"2023-11-29 14:00:00","tz":"+01:00","v":1701266400,"iso":"2023-11-29T14:00:00+01:00"},"forecast":{"daily":{"o3":[{"avg":18,"day":"2023-11-27","max":29,"min":11},{"avg":14,"day":"2023-11-28","max":20,"min":10},{"avg":20,"day":"2023-11-29","max":26,"min":14},{"avg":24,"day":"2023-11-30","max":29,"min":18},{"avg":27,"day":"2023-12-01","max":31,"min":26},{"avg":23,"day":"2023-12-02","max":31,"min":19},{"avg":19,"day":"2023-12-03","max":19,"min":18}],"pm10":[{"avg":11,"day":"2023-11-27","max":13,"min":7},{"avg":10,"day":"2023-11-28","max":10,"min":9},{"avg":9,"day":"2023-11-29","max":17,"min":3},{"avg":5,"day":"2023-11-30","max":7,"min":2},{"avg":3,"day":"2023-12-01","max":6,"min":2},{"avg":6,"day":"2023-12-02","max":7,"min":4},{"avg":7,"day":"2023-12-03","max":7,"min":7}],"pm25":[{"avg":39,"day":"2023-11-27","max":48,"min":24},{"avg":33,"day":"2023-11-28","max":36,"min":30},{"avg":27,"day":"2023-11-29","max":56,"min":10},{"avg":13,"day":"2023-11-30","max":23,"min":6},{"avg":8,"day":"2023-12-01","max":13,"min":4},{"avg":18,"day":"2023-12-02","max":23,"min":10},{"avg":23,"day":"2023-12-03","max":23,"min":21}]}},"debug":{"sync":"2023-11-30T00:10:51+09:00"}}}